### PR TITLE
Mark all scripted field queries as expensive

### PR DESCRIPTION
**Pull request by nik9000**
This will cause them to fail if you don't have permission to execute
expensive queries.

Relates to #59332

### CODE REVIEW

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java [hunk 1]
```diff
@@ -77,8 +78,17 @@ public String typeName() {
         return scriptFactory.newFactory(script.getParams(), context.lookup());
     }
 
+    private void checkAllowExpensiveQueries(QueryShardContext context) {
+        if (context.allowExpensiveQueries() == false) {
+            throw new IllegalArgumentException(
+                "queries cannot be executed against [script] fields while [" + ALLOW_EXPENSIVE_QUERIES.getKey() + "] is set to [false]."
```
Comment (javanna) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java:84]: nit: use the constant from the mapper? content type I think it is called

Comment (javanna) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java:84]: should we rather throw ElasticsearchException also, and why doesn't the context expose a check method instead that accepts for instance a message? :)

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java [hunk 2]
```diff
@@ -77,8 +78,17 @@ public String typeName() {
...
+            );
+        }
+    }
+
     @Override
     public Query existsQuery(QueryShardContext context) {
+        checkAllowExpensiveQueries(context);
```
Comment (javanna) [inline x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java:91]: maybe one day we will a base class for runtime mapper field types that does this in one place.

**Review by javanna — APPROVED** [review]
left a couple of nits, LGTM otherwise

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java:91]: probably!

File: x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java:84]: > should we rather throw ElasticsearchException also, and why doesn't the context expose a check method instead that accepts for instance a message? :)

That would probably be a good choice.

I have no idea why we were throwing ElasticsearchException when we were. It feels like this is what IAE is for.

Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java:84]: > I have no idea why we were throwing ElasticsearchException when we were. It feels like this is what IAE is for.

I figured it out - we translate all exceptions thrown when building a query into a `QueryShardException` which is always a 400.

Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java]: > why doesn't the context expose a check method instead that accepts for instance a message? :)

It really is a good idea. I took a look at it and, ironically, testing it is kind of a pain. I'll hold off on it.

Comment (javanna) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java]: shall we open issues for possible follow-up tasks that we don't want to spend time on right now?

Comment (nik9000) [reply x-pack/plugin/runtime-fields/src/main/java/org/elasticsearch/xpack/runtimefields/mapper/RuntimeKeywordMappedFieldType.java]: Let me give it another go this morning!

**Pull request merged**